runtime/gc_stats.c

```diff
@@ -106,23 +104,31 @@ void caml_orphan_alloc_stats(caml_domain_state *domain) {
 /* The "sampled stats" of a domain are a recent copy of its
    domain-local stats, accessed without synchronization and only
    updated ("sampled") during stop-the-world events -- each minor
-   collection, and on domain termination. */
+   collection, including those happening during domain termination. */
 static struct gc_stats sampled_gc_stats[Max_domains];
 
-/* Update the sampled stats for the given domain. */
-void caml_collect_gc_stats_sample(caml_domain_state* domain)
+/* Update the sampled stats for the given domain during a STW section. */
+void caml_collect_gc_stats_sample_stw(caml_domain_state* domain)
 {
   struct gc_stats* stats = &sampled_gc_stats[domain->id];
-  caml_collect_alloc_stats_sample(domain, &stats->alloc_stats);
-  caml_collect_heap_stats_sample(domain->shared_heap, &stats->heap_stats);
-}
-
-void caml_clear_gc_stats_sample(caml_domain_state *domain) {
-  struct gc_stats* stats = &sampled_gc_stats[domain->id];
-  memset(stats, 0, sizeof(*stats));
+  if (caml_domain_is_terminating()) {
```

Comment:
Good point, thanks. I rebased the PR to follow your suggestion. (I first tried to convince myself that redefining `caml_collect_gc_stats_sample` to take `(void)` and be special-cased on the current domain would be fine; but the rest of the minor GC code is careful to pass the domain state around, so it would feel bad to do this.)

I had trouble naming the `is_terminating` function parametrized on the domain. Clearly it should be called `caml_domain_is_terminating` (and the current one should be called `caml_domain_self_is_terminating` or something), but the `caml_domain_is_foo(void)` pattern is used in several places already. In any case, I named the new one `caml_domain_terminating`, which is not great but acceptable.